### PR TITLE
Add support for single replace in file contents

### PR DESCRIPTION
**Pull request by gsdefender**
Default behaviour is "replace all matches" as before (replaceAll = true is default)
If replaceAll = false, a single replace is attempted.

### CODE REVIEW

**Comment by floverfelt** [comment]
Thanks so much for adding this @gsdefender! Let me take a look later today

**Comment by gsdefender** [comment]
You're welcome!
I would like you to add @PeterGit96 as participant to this pull request if possible. Part of this contribution has been done by him.
Thanks.

**Comment by gsdefender** [comment]
Sorry @floverfelt, any news?

**Comment by floverfelt** [comment]
Sorry, I got busy with some other things, I'll add some comments now.

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojo.java [hunk 1]
```diff
@@ -119,6 +119,14 @@ public class FindAndReplaceMojo extends AbstractMojo {
   @Parameter(property = "encoding")
   private String encoding;
 
+  /**
+   * Whether the find and replace replaces all matches.
```
Comment (floverfelt) [inline src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojo.java]: I think this should be a little more descriptive, maybe:
"Whether the find and replace replaces all matches or just the first match."

Also, note that it applies to file and directory names as well, so:
"This applies to all filenames, directory names, and file contents. The first instance of a match in a file and/or directory name will be replaced. e.g. {findRegex}-dir-name-{findRegex} will become {replaceValue}-dir-name-{findRegex}"

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java [hunk 2]
```diff
@@ -158,17 +158,26 @@ private static void processFileContents(File file, Pattern findRegex, String rep
...
            BufferedWriter fileWriter = new BufferedWriter(osr)) {
 
-        Stream<String> lines = fileReader.lines();
+        boolean alreadyReplaced = false;
 
-        lines.forEach(line -> {
+        for (String line = fileReader.readLine(); line != null; line = fileReader.readLine()) {
           Matcher matcher = findRegex.matcher(line);
-          line = matcher.replaceAll(replaceValue);
+          if(matcher.find()) {
```
Comment (floverfelt) [inline src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java]: Space after this if

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java [hunk 3]
```diff
@@ -158,17 +158,26 @@ private static void processFileContents(File file, Pattern findRegex, String rep
...
+            if (replaceAll)
+              line = matcher.replaceAll(replaceValue);
```
Comment (floverfelt) [inline src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java:167]: Wrap this line with {} to keep it consistent stylistically.

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java [hunk 4]
```diff
@@ -158,17 +158,26 @@ private static void processFileContents(File file, Pattern findRegex, String rep
...
+            else {
+              if (!alreadyReplaced) {
```
Comment (floverfelt) [inline src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java]: You can combine this with the line above to be an else if statement

File: src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java [hunk 5]
```diff
@@ -504,23 +520,18 @@ public void testEverything() throws IOException, NoSuchFieldException, IllegalAc
   }
```
Comment (floverfelt) [inline src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java:1052]: You'll need to add some tests which verify the false condition of replaceAll not just the true. At minimum you should add tests for:
1. A file name with multiple matches and replaceAll set to false. Verify that only the first is replaced.
2. A directory name with multiple matches and replaceAll set to false. Verify that only the first is replaced.
3. A file with multiple matches and replaceAll set to false. Verify that only the first is replaced.

**Review by floverfelt — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojo.java (hunk 1 above)
Comment (gsdefender) [reply src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojo.java]: Could you please check the revised comment out?

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java (hunk 2 above)
Comment (gsdefender) [reply src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java]: Done

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java (hunk 3 above)
Comment (gsdefender) [reply src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java:167]: Done

File: src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java (hunk 5 above)
Comment (gsdefender) [reply src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java:1052]: @PeterGit96 added a comprehensive batch of tests

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java (hunk 4 above)
Comment (gsdefender) [reply src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/tasks/ProcessFilesTask.java]: Done.

File: src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java (hunk 5 above)
Comment (PeterGit96) [reply src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java:1052]: Add junit tests to verify that only the first match of (filename, directory name, file contents) is replaced with replaceAll set false;  Fix junit tests regex and asserts conditions with replaceAll set to true; All junit tests run without any error.

FILENAME
A file name with multiple matches and replaceAll set to false. Verify that only the first is replaced.

testFilenamesReplaceFirst
testFilenamesRecursiveReplaceFirst
testFilenamesRecursiveFileMasksReplaceFirst
testFilenamesRecursiveFileMasksWithExclusionsReplaceFirst

DIRECTORY NAME
A directory name with multiple matches and replaceAll set to false. Verify that only the first is replaced.

testDirectoryNamesReplaceFirst
testDirectoryNamesRecursiveReplaceFirst
testDirectoryNamesRecursiveExclusionsReplaceFirst

FILE CONTENTS
A file with multiple matches and replaceAll set to false. Verify that only the first is replaced.

testFileContentsReplaceFirst
testFileContentsNonStandardEncodingFindRegexReplaceFirst
testFileContentsNonStandardReplaceValueReplaceFirst
testFileContentsNonStandardFindAndReplaceValueReplaceFirst
testFileContentsRecursiveReplaceFirst
testFileContentsNotRecursiveReplaceFirst
testFileContentsRecursiveFileMasksReplaceFirst
testFileContentsRecursiveFileMasksExclusionsReplaceFirst
testEverythingReplaceFirst

File: src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojo.java (hunk 1 above)
Comment (floverfelt) [reply src/main/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojo.java]: Looks good to me, thanks! I might add more to the docs later as well.

File: src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java (hunk 5 above)
Comment (floverfelt) [reply src/test/java/io/github/floverfelt/find/and/replace/maven/plugin/FindAndReplaceMojoTest.java:1052]: Thank you!

**Comment by floverfelt** [comment]
@gsdefender @PeterGit96 do you need this feature on a given timeline/is it blocking? I'd like to add the fixes for the latest maven version and some other code cleanup/build checks but I could push a new version before then if need be.

**Comment by PeterGit96** [comment]
Hi @floverfelt, Thanks for your positive feedback. About the replacement mode selection feature (single/multiple), for a corporate project, we are currently using a forked version with this feature already implemented. If you were to add additional fixes and code cleanup/build checks, it wouldn't be blocking for our work activities. Once you publish the new version, we will reference it

**Review by floverfelt — APPROVED** [review]

**Pull request merged**